File: week3/cpp_hw_week3/Basic/Vector/Vector.cpp
#include <iostream>
#include <cmath>
#include "Vector.h"

Vector::Vector(double x, double y) {
    this->x = x;
    this->y = y;
}

Vector::~Vector() {

}

double Vector::getX() const {
    return this->x;
}

double Vector::getY() const {
    return this->y;
}

void Vector::setX(double value) {
    this->x = value;
}

void Vector::setY(double value) {
    this->y = value;
}

double Vector::len() const {
    return hypot(this->x, this->y);
}

bool Vector::operator==(const Vector& other) const {
    return this->x == other.x && this->y == other.y;
}

bool Vector::operator!=(const Vector& other) const {
    return this->x != other.x || this->y != other.y;
}

void Vector::operator+=(const Vector& other) {
    this->x += other.x;
    this->y += other.y;
}

void Vector::operator-=(const Vector& other) {
    this->x -= other.x;
    this->y -= other.y;
}

Vector Vector::operator+(const Vector& other) const {
    Vector result;

    result.x = this->x + other.x;
    result.y = this->y + other.y;
    return result;
}

Vector Vector::operator-(const Vector& other) const {
    Vector result;

    result.x = this->x - other.x;
    result.y = this->y - other.y;
    return result;
}

std::ostream& operator<<(std::ostream& out, const Vector& vector) {
    return out<< "(" << vector.getX() << ", " << vector.getY() << ")";
}

File: week3/java_week3/Basic/Vector/Vector.java
public class Vector {
    private double x;
    private double y;

    public Vector(double x, double y) {
        this.x = x;
        this.y = y;
    }

    public Vector() {
        this(0.0, 0.0);
    }

    public double getX() {
        return this.x;
    }

    public double getY() {
        return this.y;
    }

    public void setX(double value) {
        this.x = value;
    }

    public void setY(double value) {
        this.y = value;
    }

    public double len() {
        return Math.hypot(this.x, this.y);
    }

    public boolean equals(Vector other) {
        return this.x == other.x && this.y == other.y;
    }

    public void increment(Vector other) {
        this.x += other.x;
        this.y += other.y;
    }

    public void decrement(Vector other) {
        this.x -= other.x;
        this.y -= other.y;
    }

    public Vector sum(Vector other) {
        Vector res = new Vector(this.x + other.x, this.y + other.y);
        return res;
    }

    public Vector diff(Vector other) {
        Vector res = new Vector(this.x - other.x, this.y - other.y);
        return res;
    }

    public String toString() {
        return "(" + this.x + ", " + this.y + ")";
    }
}
